From 7e8dc7b3fbb4acea3f0238cad2ef851a67fad3b2 Mon Sep 17 00:00:00 2001
From: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
Date: Thu, 8 Jun 2023 21:01:57 +0530
Subject: [PATCH] ath12k: rate info is not properly updated for tx cmpl pkts

The rate info is not properly set in tx complete handler and so the tx
status is not properly updated to the mac80211, which fails updating
mesh metric calculation.

Fix the rate info by updating driver rate status from tx completion
feedback.

Below commit caused the rate info update failure for ath12k and hence
updated the rate in tx completion

commit 44fa75f207d8a106bc75e6230db61e961fdbf8a8
Author: Jonas Jelonek <jelonek.jonas@gmail.com>
Date:   Mon May 9 19:39:57 2022 +0200
    mac80211: extend current rate control tx status API
Link:
https://lore.kernel.org/r/20220509173958.1398201-2-jelonek.jonas@gmail.com

Signed-off-by: Karthikeyan Kathirvel <quic_kathirve@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index d03194737deb..f8b8d77cbf74 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -1007,6 +1007,7 @@ static void ath12k_dp_tx_complete_msdu(struct ath12k *ar,
 				       int ring)
 {
 	struct ieee80211_tx_status status = { 0 };
+	struct ieee80211_rate_status status_rate = { 0 };
 	struct ath12k_base *ab = ar->ab;
 	struct ieee80211_tx_info *info;
 	struct ath12k_skb_cb *skb_cb;
@@ -1122,7 +1123,13 @@ static void ath12k_dp_tx_complete_msdu(struct ath12k *ar,
 	status.skb = msdu;
 	status.info = info;
 	rate = arsta->last_txrate;
-	status.rates = &rate;
+
+	status_rate.rate_idx = rate;
+	status_rate.try_count = 1;
+
+	status.rates = &status_rate;
+	status.n_rates = 1;
+
 
 	if (unlikely(ath12k_debugfs_is_extd_tx_stats_enabled(ar))) {
 		if(arsta->wbm_tx_stats && wbm_status < HAL_WBM_REL_HTT_TX_COMP_STATUS_MAX)
-- 
2.38.0

